flow/flow.go

```diff
@@ -309,28 +312,66 @@ func (p *Packet) TransportFlow() (gopacket.Flow, error) {
 	return layer.TransportFlow(), nil
 }
 
+//Hash flow with custom func, ignore ports in transport level for VXLAN and Geneve
+func Hash(f gopacket.Flow, hasher hash.Hash, isTransportUDP bool) {
+	src, dst := f.Endpoints()
+	if isTransportUDP {
+		srcPort, _ := strconv.ParseInt(src.String(), 10, 64)
+		dstPort, _ := strconv.ParseInt(dst.String(), 10, 64)
+		if srcPort == 4789 || srcPort == 6081 {
+			hasher.Write(src.Raw())
+			value64 := make([]byte, 8)
+			binary.BigEndian.PutUint64(value64, uint64(f.EndpointType()))
+			hasher.Write(value64)
+			return
+		}
+		if dstPort == 4789 || dstPort == 6081 {
+			hasher.Write(dst.Raw())
+			value64 := make([]byte, 8)
+			binary.BigEndian.PutUint64(value64, uint64(f.EndpointType()))
+			hasher.Write(value64)
+			return
+		}
+	}
+	if bytes.Compare(src.Raw(), dst.Raw()) > 0 {
+		hasher.Write(src.Raw())
+		hasher.Write(dst.Raw())
+	} else {
+		hasher.Write(dst.Raw())
+		hasher.Write(src.Raw())
+	}
+
+	value64 := make([]byte, 8)
+	binary.BigEndian.PutUint64(value64, uint64(f.EndpointType()))
+	hasher.Write(value64)
+}
+
 // Key returns the unique flow key
 // The unique key is calculated based on parentUUID, network, transport and applicable layers
 func (p *Packet) Key(parentUUID string, opts Opts) string {
-	var uuid uint64
-
+	hasher := xxHash64.New(0)
 	// uses L2 is requested or if there is no network layer
 	if opts.LayerKeyMode == L2KeyMode || p.NetworkLayer() == nil {
 		if layer := p.LinkLayer(); layer != nil {
-			uuid ^= layer.LinkFlow().FastHash()
+			Hash(layer.LinkFlow(), hasher, false)
+			b := make([]byte, 8)
+			binary.BigEndian.PutUint64(b, uint64(linkID(p)))
+			hasher.Write(b)
 		}
 	}
 	if layer := p.NetworkLayer(); layer != nil {
-		uuid ^= layer.NetworkFlow().FastHash()
+		Hash(layer.NetworkFlow(), hasher, false)
+		b := make([]byte, 8)
+		binary.BigEndian.PutUint64(b, uint64(networkID(p)))
```

Comment:
Why is this needed ?

Reply:
if you drop the last line, it would result in mismatching flows (those that have Geneve application layer, because we need the VNI in order to distinguish between them)
** you cant test it yourself -- make test will fail